--- cyrus-imapd/perl/imap/Makefile.PL.in.old  2021-09-25 11:47:41.223740000 +0100
+++ cyrus-imapd/perl/imap/Makefile.PL.in      2021-09-25 11:53:25.598923100 +0100
@@ -80,6 +80,32 @@
 EOT
 }
 
+if ( defined &ExtUtils::Liblist::ext ) {
+    warn "Patching ExtUtils::Liblist::ext\n";
+    my $orig = \&ExtUtils::Liblist::ext;
+    no warnings 'redefine';
+    *ExtUtils::Liblist::ext = sub {
+        my ( @result ) = $orig->(@_);
+        if ( $result[2] !~ /@LIB_SASL@/ ) {
+            warn "Reinserting stripped @LIB_SASL@\n";
+            $result[2] .= ' @LIB_SASL@';
+        }
+        if ( $result[2] !~ /@ZLIB@/ ) {
+            warn "Reinserting stripped @ZLIB@\n";
+            $result[2] .= ' @ZLIB@';
+        }
+        if ( $result[2] !~ /@LIB_UUID@/ ) {
+            warn "Reinserting stripped @LIB_UUID@\n";
+            $result[2] .= ' @LIB_UUID@';
+        }
+        if ( $result[2] !~ /@SQLITE_LIBADD@/ ) {
+            warn "Reinserting stripped @SQLITE_LIBADD@\n";
+            $result[2] .= ' @SQLITE_LIBADD@';
+        }
+        return @result;
+    };
+}
+
 WriteMakefile(
     'NAME'	=> 'Cyrus::IMAP',
     'ABSTRACT'  => 'Cyrus administrative interface',
--- cyrus-imapd/perl/sieve/managesieve/Makefile.PL.in.old  2021-09-25 11:47:41.223740000 +0100
+++ cyrus-imapd/perl/sieve/managesieve/Makefile.PL.in      2021-09-25 11:53:25.598923100 +0100
@@ -62,6 +62,32 @@
     ;
 }

+if ( defined &ExtUtils::Liblist::ext ) {
+    warn "Patching ExtUtils::Liblist::ext\n";
+    my $orig = \&ExtUtils::Liblist::ext;
+    no warnings 'redefine';
+    *ExtUtils::Liblist::ext = sub {
+        my ( @result ) = $orig->(@_);
+        if ( $result[2] !~ /@LIB_SASL@/ ) {
+            warn "Reinserting stripped @LIB_SASL@\n";
+            $result[2] .= ' @LIB_SASL@';
+        }
+        if ( $result[2] !~ /@ZLIB@/ ) {
+            warn "Reinserting stripped @ZLIB@\n";
+            $result[2] .= ' @ZLIB@';
+        }
+        if ( $result[2] !~ /@LIB_UUID@/ ) {
+            warn "Reinserting stripped @LIB_UUID@\n";
+            $result[2] .= ' @LIB_UUID@';
+        }
+        if ( $result[2] !~ /@SQLITE_LIBADD@/ ) {
+            warn "Reinserting stripped @SQLITE_LIBADD@\n";
+            $result[2] .= ' @SQLITE_LIBADD@';
+        }
+        return @result;
+    };
+}
+
 # See lib/ExtUtils/MakeMaker.pm for details of how to influence
 # the contents of the Makefile that is written.
 WriteMakefile(
