diff --git a/pretty.cabal b/pretty.cabal
index 7a77b62..c353bd5 100644
--- a/pretty.cabal
+++ b/pretty.cabal
@@ -40,8 +40,7 @@ Library
         Text.PrettyPrint.Annotated.HughesPJ
         Text.PrettyPrint.Annotated.HughesPJClass
     build-depends: base >= 4.5 && < 5,
-                   deepseq >= 1.1,
-                   ghc-prim
+                   deepseq >= 1.1
     extensions: CPP, BangPatterns, DeriveGeneric
     ghc-options: -Wall -fwarn-tabs
 
@@ -51,7 +50,6 @@ Test-Suite test-pretty
                     src
     build-depends: base >= 4.5 && < 5,
                    deepseq >= 1.1,
-                   ghc-prim,
                    QuickCheck >= 2.5 && <3
     main-is: Test.hs
     other-modules:
diff --git a/src/Text/PrettyPrint/Annotated/HughesPJ.hs b/src/Text/PrettyPrint/Annotated/HughesPJ.hs
index 44cce28..ea0d3d6 100644
--- a/src/Text/PrettyPrint/Annotated/HughesPJ.hs
+++ b/src/Text/PrettyPrint/Annotated/HughesPJ.hs
@@ -87,7 +87,7 @@ import Data.Function   ( on )
 #if __GLASGOW_HASKELL__ >= 803
 import Prelude         hiding ( (<>) )
 #endif
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 || defined(__MHS__)
 import qualified Data.Semigroup as Semi ( Semigroup((<>)) )
 #elif __GLASGOW_HASKELL__ < 709
 import Data.Monoid     ( Monoid(mempty, mappend)  )
@@ -258,12 +258,12 @@ data TextDetails = Chr  {-# UNPACK #-} !Char -- ^ A single Char fragment
                  | PStr String -- ^ Used to represent a Fast String fragment
                                --   but now deprecated and identical to the
                                --   Str constructor.
-#if __GLASGOW_HASKELL__ >= 701
+#if __GLASGOW_HASKELL__ >= 701 || defined(__MHS__)
                  deriving (Show, Eq, Generic)
 #endif
 
 -- Combining @Doc@ values
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 || defined(__MHS__)
 instance Semi.Semigroup (Doc a) where
 #ifndef TESTING
     (<>) = (Text.PrettyPrint.Annotated.HughesPJ.<>)
diff --git a/src/Text/PrettyPrint/HughesPJ.hs b/src/Text/PrettyPrint/HughesPJ.hs
index 5d06aeb..ce7f1cc 100644
--- a/src/Text/PrettyPrint/HughesPJ.hs
+++ b/src/Text/PrettyPrint/HughesPJ.hs
@@ -89,7 +89,7 @@ import Data.Function   ( on )
 #if __GLASGOW_HASKELL__ >= 803
 import Prelude         hiding ( (<>) )
 #endif
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 || defined(__MHS__)
 import qualified Data.Semigroup as Semi ( Semigroup((<>)) )
 #elif __GLASGOW_HASKELL__ < 709
 import Data.Monoid     ( Monoid(mempty, mappend)  )
@@ -130,7 +130,7 @@ liftBinary f (Doc a) (Doc b) = Doc (f a b)
 type RDoc = Doc
 
 -- Combining @Doc@ values
-#if __GLASGOW_HASKELL__ >= 800
+#if __GLASGOW_HASKELL__ >= 800 || defined(__MHS__)
 instance Semi.Semigroup Doc where
     (<>) = (Text.PrettyPrint.HughesPJ.<>)
 
