From f9a03e5fa2d9665cba2100cc04070385a93267fd Mon Sep 17 00:00:00 2001
From: Alex Tunstall <alex@tunstall.xyz>
Date: Sun, 30 Mar 2025 10:50:06 +0100
Subject: [PATCH] Add atomicWriteIORef

---
 lib/Data/IORef.hs | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/lib/Data/IORef.hs b/lib/Data/IORef.hs
index b402fea8..3f8eb72a 100644
--- a/lib/Data/IORef.hs
+++ b/lib/Data/IORef.hs
@@ -4,6 +4,7 @@ module Data.IORef(
   readIORef,
   writeIORef,
   modifyIORef,
+  atomicWriteIORef,
   atomicModifyIORef',
   ) where
 import qualified Prelude()              -- do not import Prelude
@@ -31,6 +32,9 @@ modifyIORef :: forall a . IORef a -> (a -> a) -> IO ()
 modifyIORef (R p) f = primArrRead p 0 `primBind` \ a -> primArrWrite p 0 (f a)
 
 -- XXX Neither atomic nor strict
+atomicWriteIORef :: IORef a -> a -> IO ()
+atomicWriteIORef = writeIORef
+
 atomicModifyIORef, atomicModifyIORef' :: IORef a -> (a -> (a, b)) -> IO b
 atomicModifyIORef (R p) f = primArrRead p 0 `primBind` \a
   -> let (a', b) = f a in primArrWrite p 0 a' `primBind` \_ -> primReturn b
-- 
2.47.2

