diff --git a/cpphs-1.20.9/Language/Preprocessor/Cpphs/CppIfdef.hs b/cpphs-1.20.9/Language/Preprocessor/Cpphs/CppIfdef.hs
index 97ed942..7035419 100644
--- a/cpphssrc/malcolm-wallace-universe/cpphs-1.20.9/Language/Preprocessor/Cpphs/CppIfdef.hs
+++ b/cpphssrc/malcolm-wallace-universe/cpphs-1.20.9/Language/Preprocessor/Cpphs/CppIfdef.hs
@@ -233,7 +233,7 @@ expandSymOrCall st =
         <|> convert sym []
       )
   where
-    fragment = many1 (satisfy (`notElem`",)"))
+    fragment = many1 (satisfy (`notElem`(",)"::String)))
     convert "defined" [arg] =
       case lookupST arg st of
         Nothing | all isDigit arg    -> return arg
@@ -254,10 +254,6 @@ expandSymOrCall st =
                     if null args then return sym
                     else return "0"
                  -- else fail (disp sym args++" explicitly undefined with -U")
-    disp sym args = let len = length args
-                        chars = map (:[]) ['a'..'z']
-                    in sym ++ if null args then ""
-                              else "("++intercalate "," (take len chars)++")"
 
 parseBoolExp :: TextParser Bool
 parseBoolExp =
@@ -387,13 +383,13 @@ recursivelyExpand st inp =
     (Right s,  _) -> s
 
 parseSym :: TextParser String
-parseSym = many1 (satisfy (\c-> isAlphaNum c || c`elem`"'`_"))
+parseSym = many1 (satisfy (\c-> isAlphaNum c || c`elem`("'`_"::String)))
            `onFail`
            do xs <- allAsString
               fail $ "Expected an identifier, got \""++xs++"\""
 
 notIdent :: Char -> Bool
-notIdent c = not (isAlphaNum c || c`elem`"'`_")
+notIdent c = not (isAlphaNum c || c`elem`("'`_"::String))
 
 skip :: TextParser a -> TextParser a
 skip p = many (satisfy isSpace) >> p
diff --git a/cpphs-1.20.9/Language/Preprocessor/Cpphs/Position.hs b/cpphs-1.20.9/Language/Preprocessor/Cpphs/Position.hs
index 4a93070..24edd03 100644
--- a/cpphssrc/malcolm-wallace-universe/cpphs-1.20.9/Language/Preprocessor/Cpphs/Position.hs
+++ b/cpphssrc/malcolm-wallace-universe/cpphs-1.20.9/Language/Preprocessor/Cpphs/Position.hs
@@ -94,7 +94,7 @@ cpp2hask line | "#line" `isPrefixOf` line = "{-# LINE "
 -- | Strip non-directory suffix from file name (analogous to the shell
 --   command of the same name).
 dirname :: String -> String
-dirname  = reverse . safetail . dropWhile (not.(`elem`"\\/")) . reverse
+dirname  = reverse . safetail . dropWhile (not.(`elem`("\\/"::String))) . reverse
   where safetail [] = []
         safetail (_:x) = x
 
diff --git a/cpphs-1.20.9/Language/Preprocessor/Cpphs/Tokenise.hs b/cpphs-1.20.9/Language/Preprocessor/Cpphs/Tokenise.hs
index 405eb5f..2771034 100644
--- a/cpphssrc/malcolm-wallace-universe/cpphs-1.20.9/Language/Preprocessor/Cpphs/Tokenise.hs
+++ b/cpphssrc/malcolm-wallace-universe/cpphs-1.20.9/Language/Preprocessor/Cpphs/Tokenise.hs
@@ -244,11 +244,11 @@ tokenise stripEol stripComments ansi lang ((pos,str):pos_strs) =
   plaintext _    acc _ [] []            = emit acc $ []
 
   -- predicates for lexing Haskell.
-  ident0 x = isAlpha x    || x `elem` "_`"
-  ident1 x = isAlphaNum x || x `elem` "'_`"
-  symbol x = x `elem` ":!#$%&*+./<=>?@\\^|-~"
-  single x = x `elem` "(),[];{}"
-  space  x = x `elem` " \t"
+  ident0 x = isAlpha x    || x `elem` ("_`" :: String)
+  ident1 x = isAlphaNum x || x `elem` ("'_`" :: String)
+  symbol x = x `elem` (":!#$%&*+./<=>?@\\^|-~" :: String)
+  single x = x `elem` ("(),[];{}" :: String)
+  space  x = x `elem` (" \t" :: String)
   -- conversion of comment text to whitespace
   white '\n' = '\n'
   white '\r' = '\r'
diff --git a/cpphs-1.20.9/cpphs.hs b/cpphs-1.20.9/cpphs.hs
index 6760e6e..bf2d757 100644
--- a/cpphssrc/malcolm-wallace-universe/cpphs-1.20.9/cpphs.hs
+++ b/cpphssrc/malcolm-wallace-universe/cpphs-1.20.9/cpphs.hs
@@ -87,7 +87,7 @@ data ConvertArgs = ConvertArgs { traditional, strip :: Bool
 convertArgs :: [String] -> [String]
 convertArgs xs = f (ConvertArgs False True "-" "-") xs
     where
-        flg = "DUI"
+        flg = "DUI" :: String
 
         f e (['-',r]:x:xs) | r `elem` flg = ('-':r:x) : f e xs
         f e (x@('-':r:_):xs) | r `elem` flg = x : f e xs
diff --git a/polyparse-1.12/src/Text/Parse.hs b/polyparse-1.12/src/Text/Parse.hs
index 2d0e8b3..ec367f1 100644
--- a/cpphssrc/malcolm-wallace-universe/polyparse-1.12/src/Text/Parse.hs
+++ b/cpphssrc/malcolm-wallace-universe/polyparse-1.12/src/Text/Parse.hs
@@ -128,15 +128,15 @@ word = P p
             | isDigit  c = let (ds,t)  = span isDigit s in
                            lexFracExp (c:ds) t
             | otherwise  = Failure (c:s) ("Bad character: "++show c)
-             where isSingle c  =  c `elem` ",;()[]{}`"
-                   isSym    c  =  c `elem` "!@#$%&*+./<=>?\\^|:-~"
+             where isSingle c  =  c `elem` (",;()[]{}`" :: String)
+                   isSym    c  =  c `elem` ("!@#$%&*+./<=>?\\^|:-~" :: String)
                    isIdInit c  =  isAlpha c || c == '_'
-                   isIdChar c  =  isAlphaNum c || c `elem` "_'"
+                   isIdChar c  =  isAlphaNum c || c `elem` ("_'" :: String)
                    lexFracExp acc ('.':d:s) | isDigit d   =
                                       lexExp (acc++'.':d:ds) t
                                               where (ds,t) = span isDigit s
                    lexFracExp acc s = lexExp acc s
-                   lexExp     acc (e:s) | e`elem`"eE" =
+                   lexExp     acc (e:s) | e`elem`("eE" :: String) =
                                       case s of
                                         ('+':d:t) | isDigit d ->
                                                     let (ds,u)=span isDigit t in
